Update to docker 1.13.x

vdemeester opened the pull request:
###### Motivation for this change

`docker` 1.13.0 is going to be released next month (more or less). This shouldn't be merged before official release but gives time to review. This change the package quite a bit, so let me describe the changes :

- There is more binaries now : `docker`, `dockerd`, `docker-proxy`, `docker-runc`, `docker-containerd`, `docker-containerd-shim` and `docker-init` — from 5 projects : `docker/docker`, `docker/runc` (to follow docker release), `docker/containerd`, `docker-libnetwork` and `krallin/tini`.
- The package builds all the pieces itself instead of depending to other packages like `runc` and `containerd`. This makes sure we have the same binary (from the same git commit) than the official release and thus have no weird bugs because of version mismatch.
- The man pages of `docker` are in the process of beeing automatically generate so there is some missing man page in the current package. This aims to fix that as well (*still working on it*).
- This adds a `experimental` options to the `docker` service too – to start `docker` with experimental features (using `--experimental`).

There is still a few things to do :
- <del>[ ] Fix missing man pages</del> Currently, part of the manpages are generated in a docker container. I have no clean way to build manpage otherwise for now.. 😓 
- [ ] Handle version and build time better (`docker version` and `docker info`) 
  ```
   containerd version:  (expected: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e)
   runc version: N/A (expected: 51371867a01c467f08af739783b8beafc154c4d7)
   init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
  ```
- [x] Update the package to `1.13.0` when released (and on each rc too)

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---


- Update docker version to 1.13
- Build containerd, runc, tini, docker-proxy from the right
  version (little duplicated from conf in docker scripts)

vdemeester commented on pkgs/applications/virtualization/docker/default.nix:
```diff
@@ -1,78 +1,156 @@
     ++ optional (devicemapper == null) "exclude_graphdriver_devicemapper";
 
+    postUnpack = ''
+    cp -fR ${dockerSrc} docker
```
I did get errors with chmod… I'm not used to multiple sources build so I *hacked* around for that… There might be a better solution 👼

vdemeester commented on pkgs/applications/virtualization/docker/default.nix:
```diff
@@ -1,78 +1,156 @@
-, go, containerd, runc
-, sqlite, iproute, bridge-utils, devicemapper, systemd
+{ stdenv, lib, fetchFromGitHub, makeWrapper, cmake, pkgconfig
+, go-md2man, libapparmor, apparmor-parser, libseccomp, git
```
runc needs git (for the commit hash I guess). I really wish there was another way like what we do for docker/docker but it doesn't seem so…

offlinehacker commented:
But we have separated packages for runs and containers, why did you put them all into docker nix expression?

offlinehacker commented:
You can update run and containerd, and use latest stable release. If docker's release process is bad, it does not mean we have to do equal.

vdemeester commented:
> But we have separated packages for runs and containers, why did you put them all into docker nix expression?

So about that, there is a few reasons why I made that change.

For each `docker` release, it depends on specific version of `containerd`, `runc`, etc.. If you're not running the right version of it, you might encounter weird bugs and it might be tricky to debug or get help/support because of that.

Although package `containerd` and `runc` on their own is a good idea — because they can be used without `docker` — I'm don't really like the way it's currently done :
- `containerd` and `runc` are *copied* in `docker-runc` and `docker-containerd`/`docker-containerd-shim` and packaged in the final `docker` package — this feels weird to me. Going on that path, it would make more sence to not do that (and not including `docker-runc` and `docker-containerd*` in the package), and use daemon flags to tell `docker` where to find the runtime and the containerd socket).
- (*I might be wrong on that one*) as far as I can tell, I cannot *pin* a certain version of `runc` and `containerd` for this specific package. Let say I update `runc` and `containerd` to the right version (and I think I will, probably in separate PRs) and I depend on it. As `containerd` and `runc` are separate project, they could be released differently (most often, …). They could (sholud?) be updated separately from `docker` in nixpkgs (especially `runc` which is an oci project).  And thus, I have no guarantee that the `docker-runc` and `docker-containerd` for this specific version of `docker` will stay the same and the one that work and should be shipped with `docker`. And then again, you could run into really weird bugs.

> You can update runc and containerd, and use latest stable release. If docker's release process is bad, it does not mean we have to do equal.

I don't think there is good or bad release process. See this (packaging) as an embedded use of `runc`, `containerd`, etc..

offlinehacker commented:
Please look on GitHub, I have rephrased my comment a bit, first it was
written in a rush(and GitHub does not send emails for updates). Sorry for
that.

On Thu, Nov 24, 2016, 10:47 AM Vincent Demeester <notifications@github.com>
wrote:

> But we have separated packages for runs and containers, why did you put
> them all into docker nix expression?
>
> So about that, there is a few reasons why I made that change.
>
> For each docker release, it depends on specific version of containerd,
> runc, etc.. If you're not running the right version of it, you might
> encounter weird bugs and it might be tricky to debug or get help/support
> because of that.
>
> Although package containerd and runc on their own is a good idea —
> because they can be used without docker — I'm don't really like the way
> it's currently done :
>
>    - containerd and runc are *copied* in docker-runc and docker-containerd
>    /docker-containerd-shim and packaged in the final docker package —
>    this feels weird to me. Going on that path, it would make more sence to not
>    do that (and not including docker-runc and docker-containerd* in the
>    package), and use daemon flags to tell docker where to find the
>    runtime and the containerd socket).
>    - (*I might be wrong on that one*) as far as I can tell, I cannot *pin*
>    a certain version of runc and containerd for this specific package.
>    Let say I update runc and containerd to the right version (and I think
>    I will, probably in separate PRs) and I depend on it. As containerd
>    and runc are separate project, they could be released differently
>    (most often, …). They could (sholud?) be updated separately from docker
>    in nixpkgs (especially runc which is an oci project). And thus, I have
>    no guarantee that the docker-runc and docker-containerd for this
>    specific version of docker will stay the same and the one that work
>    and should be shipped with docker. And then again, you could run into
>    really weird bugs.
>
> You can update runc and containerd, and use latest stable release. If
> docker's release process is bad, it does not mean we have to do equal.
>
> I don't think there is good or bad release process. See this (packaging)
> as an embedded use of runc, containerd, etc..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/20656#issuecomment-262733466>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAjvS6kQ-QabuIDTrW8a46wbcyHtNHIPks5rBV0ygaJpZM4K6zkA>
> .
>

offlinehacker commented:
The one other thing I forgot to mention, if overriding derivations does not
work(look overrideDerivation), we can ship multiple versions of runc and
containerd(copy expressions, with a new name, expose in all-packages.nix),
and name them differently like we have for many packages(nodejs as one
example). We could name them for example runc-docker and containerd-docker.

On Thu, Nov 24, 2016, 10:45 PM Jaka Hudoklin <jakahudoklin@gmail.com> wrote:

> Please look on GitHub, I have rephrased my comment a bit, first it was
> written in a rush(and GitHub does not send emails for updates). Sorry for
> that.
>
> On Thu, Nov 24, 2016, 10:47 AM Vincent Demeester <notifications@github.com>
> wrote:
>
> But we have separated packages for runs and containers, why did you put
> them all into docker nix expression?
>
> So about that, there is a few reasons why I made that change.
>
> For each docker release, it depends on specific version of containerd,
> runc, etc.. If you're not running the right version of it, you might
> encounter weird bugs and it might be tricky to debug or get help/support
> because of that.
>
> Although package containerd and runc on their own is a good idea —
> because they can be used without docker — I'm don't really like the way
> it's currently done :
>
>    - containerd and runc are *copied* in docker-runc and docker-containerd
>    /docker-containerd-shim and packaged in the final docker package —
>    this feels weird to me. Going on that path, it would make more sence to not
>    do that (and not including docker-runc and docker-containerd* in the
>    package), and use daemon flags to tell docker where to find the
>    runtime and the containerd socket).
>    - (*I might be wrong on that one*) as far as I can tell, I cannot *pin*
>    a certain version of runc and containerd for this specific package.
>    Let say I update runc and containerd to the right version (and I think
>    I will, probably in separate PRs) and I depend on it. As containerd
>    and runc are separate project, they could be released differently
>    (most often, …). They could (sholud?) be updated separately from docker
>    in nixpkgs (especially runc which is an oci project). And thus, I have
>    no guarantee that the docker-runc and docker-containerd for this
>    specific version of docker will stay the same and the one that work
>    and should be shipped with docker. And then again, you could run into
>    really weird bugs.
>
> You can update runc and containerd, and use latest stable release. If
> docker's release process is bad, it does not mean we have to do equal.
>
> I don't think there is good or bad release process. See this (packaging)
> as an embedded use of runc, containerd, etc..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/20656#issuecomment-262733466>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAjvS6kQ-QabuIDTrW8a46wbcyHtNHIPks5rBV0ygaJpZM4K6zkA>
> .
>
>

vdemeester commented:
@offlinehacker ah, I think I see what you mean. I'll try to update accordingly soon 😉

vdemeester commented:
Alright, RC3 is out, I'll update this :angel:

vdemeester commented:
Updated with `1.13.0-rc3` and using `overrideDerivation` instead, introducing `docker-proxy`.
I also update `tini` and make sure we can build `tini-static`. I'm not sure it's the right way to do it though ?

offlinehacker commented:
Besides comnents above, looks good to me :)

vdemeester commented on pkgs/applications/virtualization/docker/proxy.nix:
```diff
@@ -0,0 +1,36 @@
+
+with lib;
+
+stdenv.mkDerivation rec {
```
Oh right, I'll do that 👼

vdemeester commented on pkgs/applications/virtualization/docker/proxy.nix:
```diff
@@ -0,0 +1,36 @@
+  '';
+
+  meta = {
+    description = "Docker garbage collection of containers and images";
```
Oh wow, I guess I was a little bit tired 😂

vdemeester commented on pkgs/applications/virtualization/docker/default.nix:
```diff
@@ -11,15 +11,50 @@ with lib;
+      sha256 = "0zj4kdis1vvc6dwn4gplqna0bs7v6d1y2zc8v80s3zi018inhznw";
+    };
+
+    patchPhase = ''
```
In `tini` the `cmake` file is patch to remove `static` from it. But in the case of docker, `tini` (`docker-init` in the end) have to be static. This is to override the `patchPhase` — wasn't sure it was the right move to it though.

vdemeester commented:
@offlinehacker thanks 👍 I still need to fix the manpages (as they are incompletely generate for now) and then we'll just have to wait for the release 👼

offlinehacker commented:
Nice :)

vdemeester commented:
Updated to 1.13.0-rc4… I'll update for 1.13.0-rc5 this week.
Man page is gonna be a little bit tricky 😅 Gonna need some `go` library in build input as part of the man are generated by a go program..

Mic92 commented:
@vdemeester don't they use a command line application to compile documentation from markdown? You can probably add this as a separate package.

aneeshusa commented:
[`overrideAttrs`](https://github.com/NixOS/nixpkgs/pull/18660) is a new alternative to `overrideDerivation` which is now preferred to use.

vdemeester committed:
docker: 1.12.6 -> 1.13.0

- Update docker version to 1.13.0.
- Introduce now docker-proxy package (from libnetmork).
- Use overrideDerivation to set the correct version for docker.
- Update tini to make sure we can build it static.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

phanimahesh commented:
docker 1.13 has been released few days ago. I just started locally patching the derivation and found this when I was halfway.

@vdemeester Will you be updatingthis to use `buildGoPackage` instead of `mkDerivation`?
@offlinehacker Any further comments or is this LGTM?

offlinehacker commented:
LGTM
